### PR TITLE
feat: Create Hacker mode theme #1988

### LINKED ISSUE #1988: "Hacker mode" theme
### Check for duplicates

- [X] I have searched for similar issues before opening a new one.

### Component

New theme plugin.

### Problem

I think code looks cooler in black and green.

### Request

Create a theme with:
- black background
- black block fill for all block types
- white outline for all block types
- green text for all block types
- dark toolbox
- dark flyout
- monospace text

Starter image of what I'm thinking about:
image

### Alternatives considered

_No response_

### Additional context

An all-white mode with monospace text would also be fun.

## To Fix

1) Run the `@blockly/create-package` script to create a new theme plugin.
2) Create the theme. You can check out our [theme docs](https://developers.google.com/blockly/guides/configure/web/appearance/themes) for help.
3) Run `npm run start` to test that your theme looks correct.

### PR DESCRIPTION
**Pull request by shivalipatel6**
I created a theme with a black background, black block fill for all blocks, a white shadow for all blocks (still working on putting in an actual outline), dark toolbox, dark flyout, and monospace text. I am still working on putting in green text for all block text. Need to fix CSS issue first.

## The basics

- [ X] I [validated my changes](https://developers.google.com/blockly/guides/contribute/samples#making_and_verifying_a_change)

### Resolves

Fixes #1988

### Proposed Changes
This pull request creates a new blockly theme that has a black background for everything, black fill in for blocks and a white highlight. Theme is currently incomplete as it is missing green text color, but it will be fixed later once some CSS work is fixed. 
hacker-modescreenshot1

### Documentation

 Once the theme is made public, the documentation about themes will need to be updated to let users know the theme exists. As well documentation about how to download the theme will be needed.

### CODE REVIEW

File: plugins/theme-hackermode/package.json [hunk 1]
```diff
@@ -0,0 +1,54 @@
+{
```
Comment (BeksOmega) [inline plugins/theme-hackermode/package.json:1]: Can you mark this package as [private](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#private)?

File: plugins/theme-hackermode/package.json
Comment (BeksOmega) [inline plugins/theme-hackermode/package.json:1]: Sorry the create-package script doesn't work quite correctly for first party plugins, so this file needs some manual fixes. Will list below!

File: plugins/theme-hackermode/package.json [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  "name": "blockly-theme-hackermode",
```
Comment (BeksOmega) [inline plugins/theme-hackermode/package.json]:
```suggestion
  "name": "@blockly/theme-hackermode",
```

File: plugins/theme-hackermode/package.json [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  "version": "0.0.0",
+  "description": "A Blockly theme.",
```
Comment (BeksOmega) [inline plugins/theme-hackermode/package.json]: Can you update this to be more descriptive?

File: plugins/theme-hackermode/README.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
+# blockly-theme-hackermode [![Built on Blockly](https://tinyurl.com/built-on-blockly)](https://github.com/google/blockly)
```
Comment (BeksOmega) [inline plugins/theme-hackermode/README.md]:
```suggestion
# @blockly/theme-hackermode [![Built on Blockly](https://tinyurl.com/built-on-blockly)](https://github.com/google/blockly)
```

File: plugins/theme-hackermode/README.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+```
+npm install blockly-theme-hackermode --save
+```
+
+## Usage
+
+```js
+import * as Blockly from 'blockly';
+import Theme from 'blockly-theme-hackermode';
```
Comment (BeksOmega) [inline plugins/theme-hackermode/README.md]:
```suggestion
import Theme from '@blockly/theme-hackermode';
```

File: plugins/theme-hackermode/README.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
+# blockly-theme-hackermode [![Built on Blockly](https://tinyurl.com/built-on-blockly)](https://github.com/google/blockly)
+
+<!--
+  - TODO: Add theme description.
+  -->
+
+A [Blockly](https://www.npmjs.com/package/blockly) theme that ...
```
Comment (BeksOmega) [inline plugins/theme-hackermode/README.md]: Can you update this to be more descriptive?

File: plugins/theme-hackermode/README.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+## Installation
+
+### Yarn
+
+```
+yarn add blockly-theme-hackermode
```
Comment (BeksOmega) [inline plugins/theme-hackermode/README.md]:
```suggestion
yarn add @blockly/theme-hackermode
```

File: plugins/theme-hackermode/README.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+```
+
+### npm
+
+```
+npm install blockly-theme-hackermode --save
```
Comment (BeksOmega) [inline plugins/theme-hackermode/README.md]:
```suggestion
npm install @blockly/theme-hackermode --save
```

File: plugins/theme-hackermode/GETSTARTED.md
Comment (BeksOmega) [inline plugins/theme-hackermode/GETSTARTED.md:1]: Please delete this file.

File: plugins/theme-hackermode/package.json [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+  ],
+  "devDependencies": {
+    "blockly": "^10.2.2",
+    "@blockly/dev-scripts": "^3.1.0",
+    "@blockly/dev-tools": "^7.1.3"
+  },
+  "peerDependencies": {
+    "blockly": "^10.2.2"
+  },
+  "publishConfig": {},
```
Comment (BeksOmega) [inline plugins/theme-hackermode/package.json]:
```suggestion
  "publishConfig": {
    "access": "public",
    "registry": "https://wombat-dressing-room.appspot.com"
  },
```

File: plugins/theme-hackermode/src/index.js [hunk 10]
```diff
@@ -0,0 +1,116 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// TODO: Edit theme overview.
+/**
+ * @fileoverview Theme overview.
+ */
```
Comment (BeksOmega) [inline plugins/theme-hackermode/src/index.js]: Please complete.

File: plugins/theme-hackermode/src/index.js [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+import Blockly from 'blockly/core';
```
Comment (BeksOmega) [inline plugins/theme-hackermode/src/index.js]:
```suggestion
import * as Blockly from 'blockly/core';
```

File: plugins/theme-hackermode/package.json [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  "module": "./src/index.js",
+  "unpkg": "./dist/index.js",
+  "author": "",
+  "keywords": [
+    "blockly",
+    "blockly-plugin",
+    "blockly-theme",
+    "theme-hackermode"
+  ],
+  "homepage": "git@github.com:shivalipatel6/blockly-samples/tree/master/plugins/theme-hackermode#readme",
```
Comment (BeksOmega) [inline plugins/theme-hackermode/package.json]:
```suggestion
  "homepage": "git@github.com:google/blockly-samples/tree/master/plugins/theme-hackermode#readme",
```

File: plugins/theme-hackermode/package.json [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+  "bugs": {
+    "url": "git@github.com:shivalipatel6/blockly-samples/issues"
```
Comment (BeksOmega) [inline plugins/theme-hackermode/package.json]:
```suggestion
    "url": "git@github.com:google/blockly-samples/issues"
```

File: plugins/theme-hackermode/package.json [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+  },
+  "repository": {
+    "type": "git",
+    "url": "git@github.com:shivalipatel6/blockly-samples.git",
```
Comment (BeksOmega) [inline plugins/theme-hackermode/package.json]:
```suggestion
    "url": "git@github.com:google/blockly-samples.git",
```

File: plugins/theme-hackermode/src/index.js [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+  'variable_category': {
+    'colour': '#0c0d0d',
+  },
+  'variable_dynamic_category': {
+    'colour': '#0c0d0d',
+  },
+};
+
+// TODO: Edit theme name.
+export default Blockly.Theme.defineTheme('THEME_NAME', {
```
Comment (BeksOmega) [inline plugins/theme-hackermode/src/index.js]: Please complete.

File: plugins/theme-hackermode/src/index.js
Comment (BeksOmega) [inline plugins/theme-hackermode/src/index.js:1]: This file looks great =)

**Review by BeksOmega — CHANGES_REQUESTED** [review]
This is looking great so far! Thank you for putting the work into this and dealing with all of the crazy rebasing =)

**Comment by maribethb** [comment]
Hi @shivalipatel6 you can also run the create-package script with the `--first-party` flag and it will use the correct values for many of the things Beka noted. Whether you want to re-run the script and then re-apply your changes, or just fix the issues Beka noted manually is up to you!

File: plugins/theme-hackermode/src/index.js
Comment (shivalipatel6) [reply plugins/theme-hackermode/src/index.js:1]: I tackled the issues you pointed out manually. Hopefully I did everything right. Thank you so much for helping me tackle this issue and for your patience with all the problems I kept stumbling into.

File: plugins/theme-hackermode/src/index.js [hunk 16]
```diff
@@ -0,0 +1,118 @@
+private:true
```
Comment (BeksOmega) [inline plugins/theme-hackermode/src/index.js]: This needs to be added to the `package.json` file as a config property!

E.g.:

```json
  "main": "./dist/index.js",
  "module": "./src/index.js",
  "unpkg": "./dist/index.js",
  // This added.
  "private": true,
  "author": "",
```

File: plugins/theme-hackermode/src/index.js [hunk 17]
```diff
@@ -0,0 +1,118 @@
+private:true 
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+// TODO: Edit theme overview.
```
Comment (BeksOmega) [inline plugins/theme-hackermode/src/index.js]: Can you remove this TODO now that it's completed?

File: plugins/theme-hackermode/src/index.js [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+  },
+  'variable_category': {
+    'colour': '#0c0d0d',
+  },
+  'variable_dynamic_category': {
+    'colour': '#0c0d0d',
+  },
+};
+
+// TODO: Edit theme name.
```
Comment (BeksOmega) [inline plugins/theme-hackermode/src/index.js]: Can you remove this TODO?

File: plugins/theme-hackermode/README.md [hunk 19]
```diff
@@ -0,0 +1,35 @@
+# @blockly/theme-hackermode [![Built on Blockly](https://tinyurl.com/built-on-blockly)](https://github.com/google/blockly)
+<!--
+  - TODO: Add theme description.
+  -->
```
Comment (BeksOmega) [inline plugins/theme-hackermode/README.md]: Can you remove this TODO now that it's completed?

**Review by BeksOmega — CHANGES_REQUESTED** [review]
Thanks for the fixes! Just a few more nits then this should be good to go =)

**Comment by BeksOmega** [comment]
@shivalipatel6 Please resolve the build errors. You should be able to run `npm run build` to reproduce the problem. Let me know if you need help!

**Comment by shivalipatel6** [comment]
I fixed up the build issues, but I am still unsure what is up with the format issues... I could use some pointers for figuring this out.

**Comment by BeksOmega** [comment]
> I fixed up the build issues, but I am still unsure what is up with the format issues... I could use some pointers for figuring this out.

Can you try running `npm ci` and then `npm run format` from the root level of the repository? That should create some changes you can commit to fix it!

**Comment by shivalipatel6** [comment]
> > I fixed up the build issues, but I am still unsure what is up with the format issues... I could use some pointers for figuring this out.
> 
> Can you try running `npm ci` and then `npm run format` from the root level of the repository? That should create some changes you can commit to fix it!

I was able to run both commands and produce changes, but it was unable to fix the format issues. Is there anything I can run that will help me diagonse these problems?

**Comment by BeksOmega** [comment]
> > > I fixed up the build issues, but I am still unsure what is up with the format issues... I could use some pointers for figuring this out.
> > 
> > 
> > Can you try running `npm ci` and then `npm run format` from the root level of the repository? That should create some changes you can commit to fix it!
> 
> I was able to run both commands and produce changes, but it was unable to fix the format issues. Is there anything I can run that will help me diagonse these problems?

Hm you're right. I also pulled your branch and the format check passed.

Could you try rebasing onto the latest master and then rerunning the formatting commands? It could be that the quoting change you fixed in your other PR (thank you again!) is causing conflicts.

**Comment by shivalipatel6** [comment]
I ran npm ci and npm run format after I rebased and merged the changes from master into this branch but there was no changes to commit...

**Comment by BeksOmega** [comment]
Hiya @shivalipatel6 Wanted to let you know I saw this but my team is doing a hackathon this week! So I'll get back to you about this in a few days. Sorry for the delay!

**Comment by shivalipatel6** [comment]
> Hiya @shivalipatel6 Wanted to let you know I saw this but my team is doing a hackathon this week! So I'll get back to you about this in a few days. Sorry for the delay!

Have fun at the hackathon!! Good luck with your projects!

**Comment by BeksOmega** [comment]
I pushed a commit to your branch with the formatting fixes! Thanks for the work on this @shivalipatel6 =)

**Pull request merged**